builder/azure/common/client/config.go

```diff
@@ -102,12 +104,73 @@ func (c *Config) SetDefaultValues() error {
 		c.CloudEnvironmentName = DefaultCloudEnvironmentName
 	}
 
+	err := c.setNewCloudEnvironment()
+	if err != nil {
+		return err
+	}
 	return c.setCloudEnvironment()
 }
 
 func (c *Config) CloudEnvironment() *azure.Environment {
 	return c.cloudEnvironment
 }
+func (c *Config) NewCloudEnvironment() *environments.Environment {
+	return c.newCloudEnvironment
+}
+func (c *Config) AuthType() string {
+	return c.authType
+}
+
+func (c *Config) setNewCloudEnvironment() error {
+	if c.MetadataHost == "" {
+		if v := os.Getenv("ARM_METADATA_URL"); v != "" {
+			c.MetadataHost = v
+		}
+	}
+	env, err := environments.FromEndpoint(context.TODO(), c.MetadataHost, c.CloudEnvironmentName)
+	c.newCloudEnvironment = env
+	if err != nil {
+		// fall back to old method of normalizing and looking up cloud names.
+		log.Printf(fmt.Sprintf("Error looking up environment using metadata host: %s. \n"+
+			"Falling back to hardcoded mechanism...", err.Error()))
+		lookup := map[string]string{
+			"CHINA":           "china",
+			"CHINACLOUD":      "china",
+			"AZURECHINACLOUD": "china",
+
+			// TODO Implement AzureGermanCloud
```

Comment:
FYI Azure Germany doesn't exist anymore since October 2021. There are Azure regions in Germany but they are part of the general public cloud. Ref: [Azure Germany](https://learn.microsoft.com/en-us/previous-versions/azure/germany/germany-welcome)